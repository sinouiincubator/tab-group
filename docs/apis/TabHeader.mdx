---
name: TabHeader
route: /api-tab-header
menu: API
---

import { Playground, Props } from 'docz';
import TabHeader from '../../src/components/tab-header/TabHeader';

# TabHeader

<Props of={TabHeader} />
