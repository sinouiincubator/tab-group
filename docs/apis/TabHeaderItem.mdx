---
name: TabHeaderItem
route: /api-tab-header-item
menu: API
---

import { Playground, Props } from 'docz';
import TabHeaderItem from '../../src/components/tab-header-item/TabHeaderItem';

# TabHeaderItem

<Props of={TabHeaderItem} />
