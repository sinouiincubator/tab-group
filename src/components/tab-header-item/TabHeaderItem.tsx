import React, { useContext, useCallback } from 'react';
import classNames from 'classnames';
import { useRipple } from '@sinoui/ripple';
import {
  LEFT_KEY_CODE,
  TOP_KEY_CODE,
  RIGHT_KEY_CODE,
  BOTTOM_KEY_CODE,
} from 'src/constants';
import TabHeaderItemWrapper from './TabHeaderItemWrapper';
import TabListContext from '../TabListContext';
import useTabRegister from '../commons/useTabRegister';
import TabHeaderContext from '../TabHeaderContext';

interface Props {
  /**
   * 标签标题
   */
  title: React.ReactNode;
  /**
   * 指定自定义样式名
   */
  className?: string;
  /**
   * 指定自定义样式
   */
  style?: React.CSSProperties;
  /**
   * 是否可用
   */
  disabled?: boolean;
  /**
   * 指定标签页id
   */
  id?: string;
}

/**
 * 头部选项卡项组件
 */
function TabHeaderItem({ title, className, style, disabled, ...rest }: Props) {
  const tabListContext = useContext(TabListContext);
  const tabHeaderContext = useContext(TabHeaderContext);
  const index = useTabRegister();
  const selectedIndex = tabListContext ? tabListContext.selectedIndex : -1;
  const isActive = index === selectedIndex;

  const rippleRef = useRipple<HTMLDivElement>();

  const handleClick = useCallback(
    (event: React.MouseEvent) => {
      if (tabHeaderContext) {
        tabHeaderContext.onSelect(index, event);
      }
    },
    [index, tabHeaderContext],
  );

  const handleKeyDown = useCallback(
    (event: React.KeyboardEvent) => {
      const { keyCode } = event;
      if (tabListContext && tabHeaderContext) {
        const { getTabsCount } = tabListContext;
        const tabsCount = getTabsCount();
        if (keyCode === RIGHT_KEY_CODE || keyCode === BOTTOM_KEY_CODE) {
          event.stopPropagation();
          event.preventDefault();
          if (selectedIndex < tabsCount - 1) {
            tabHeaderContext.onSelect(selectedIndex + 1, event);
          } else {
            tabHeaderContext.onSelect(0, event);
          }
        } else if (keyCode === LEFT_KEY_CODE || keyCode === TOP_KEY_CODE) {
          event.stopPropagation();
          event.preventDefault();
          if (selectedIndex > 0) {
            tabHeaderContext.onSelect(selectedIndex - 1, event);
          } else {
            tabHeaderContext.onSelect(tabsCount - 1, event);
          }
        }
      }
    },
    [selectedIndex, tabHeaderContext, tabListContext],
  );

  if (index === -1) {
    return null;
  }

  return (
    <TabHeaderItemWrapper
      className={classNames('sinoui-tab-label', className, {
        'sinoui-tab-label-active': isActive,
      })}
      style={style}
      disabled={disabled}
      active={isActive}
      onClick={disabled ? undefined : handleClick}
      ref={rippleRef}
      data-testid={`tab-header-item-${index}`}
      role="tab"
      tabIndex={isActive ? 0 : -1}
      aria-disabled={disabled}
      aria-selected={isActive}
      onKeyDown={disabled ? undefined : handleKeyDown}
      {...rest}
    >
      <div className="sinoui-tab-label-content">{title}</div>
    </TabHeaderItemWrapper>
  );
}

export default TabHeaderItem;
