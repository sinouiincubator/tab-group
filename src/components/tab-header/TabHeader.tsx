import React, { useMemo } from 'react';
import TabHeaderContext from '../TabHeaderContext';
import useRefValue from '../../helpers/useRefValue';
import InnerTabHeader from './InnerTabHeader';
import TabListContext from '../TabListContext';
import useTabList from '../commons/useTabList';

interface Props {
  children?: React.ReactNode;
  /**
   * 指定在密集模式下使用。
   */
  dense?: boolean;
  /**
   * 设置当前选中的标签页。这是一个从 `0` 开始的索引，第一个标签页的索引为 `0`，第二个标签页的索引为 `1`，……
   */
  selectedIndex: number;
  /**
   * 每次标签页切换时调用的事件处理器。这个函数的 `index` 参数是新的选中标签页索引，`lastIndex` 参数是变更之前选中的标签页索引，`event` 参数是引起页签切换的事件，可能是 `keydown` 或者 `click` 事件。如果 `index` 和 `lastIndex` 相同时，表示用户在当前选中的标签页上点击。
   */
  onSelect?: (
    index: number,
    lastIndex: number,
    event: React.MouseEvent | React.KeyboardEvent,
  ) => boolean | undefined | void;
  /**
   * 在标签右侧添加附加内容。
   */
  extraContent?: React.ReactNode;
  /**
   * 设置不显示底部线条。默认为 `false`。
   */
  borderless?: boolean;

  /**
   * 设置自定义的 css 类名
   */
  className?: string;

  /**
   * 设置自定义的 css 样式
   */
  style?: React.CSSProperties;
}

/**
 * 选项卡头部组件
 *
 */
function TabHeader({
  children,
  dense,
  selectedIndex = 0,
  onSelect,
  extraContent,
  borderless,
  ...rest
}: Props) {
  const tabList = useTabList(selectedIndex);

  const onSelectRef = useRefValue(onSelect);

  const context = useMemo(
    () => ({
      dense,
      onSelect: (
        tabIndex: number,
        event: React.MouseEvent | React.KeyboardEvent,
      ) => {
        if (onSelectRef.current) {
          onSelectRef.current(tabIndex, selectedIndex, event);
        }
      },
    }),
    [dense, onSelectRef, selectedIndex],
  );

  return (
    <TabListContext.Provider value={tabList}>
      <TabHeaderContext.Provider value={context}>
        <InnerTabHeader
          extraContent={extraContent}
          borderless={borderless}
          tabListContextState={tabList}
          {...rest}
        >
          {children}
        </InnerTabHeader>
      </TabHeaderContext.Provider>
    </TabListContext.Provider>
  );
}

export default TabHeader;
