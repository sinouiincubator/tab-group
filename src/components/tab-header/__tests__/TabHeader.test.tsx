/* eslint-disable @typescript-eslint/no-non-null-assertion */
import React, { useState } from 'react';
import renderer from 'react-test-renderer';
import { ThemeProvider } from 'styled-components';
import defaultTheme from 'sinoui-components/styles/defaultTheme';
import 'jest-styled-components';
import '@testing-library/jest-dom/extend-expect';
import { render, cleanup, fireEvent } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import TabHeader from '../TabHeader';
import TabHeaderItem from '../../tab-header-item';

afterEach(cleanup);

jest.useFakeTimers();

it('正确渲染TabHeader', () => {
  const tree = renderer
    .create(
      <ThemeProvider theme={defaultTheme}>
        <TabHeader selectedIndex={0}>
          <div className="sinoui-tab-label">标签内容1</div>
          <div className="sinoui-tab-label sinoui-tab-label-active">
            标签内容2
          </div>
        </TabHeader>
      </ThemeProvider>,
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('附件内容', () => {
  const tree = renderer.create(
    <ThemeProvider theme={defaultTheme}>
      <TabHeader extraContent={<div>附件内容</div>} selectedIndex={0}>
        <div className="sinoui-tab-label">标签内容1</div>
        <div className="sinoui-tab-label sinoui-tab-label-active">
          标签内容2
        </div>
      </TabHeader>
    </ThemeProvider>,
  );

  expect(tree).toMatchSnapshot();
});

let clientRectFn: jest.Mock | undefined;
beforeEach(() => {
  clientRectFn = jest.fn();
  Element.prototype.getBoundingClientRect = clientRectFn;
});

it('inkBar显示在正确的位置上', () => {
  clientRectFn!.mockReturnValueOnce({ width: 100, left: 180 });
  clientRectFn!.mockReturnValueOnce({ left: 20 });

  const { getByTestId } = render(
    <ThemeProvider theme={defaultTheme}>
      <TabHeader selectedIndex={0}>
        <div className="sinoui-tab-label">标签内容1</div>
        <div className="sinoui-tab-label sinoui-tab-label-active">
          标签内容2
        </div>
      </TabHeader>
    </ThemeProvider>,
  );

  const inkbar = getByTestId('inkbar');

  expect(inkbar).toHaveStyle('width: 100px');
  expect(inkbar).toHaveStyle('transform: translate3d(160px, 0px, 0px)');
});

/* 验收测试 */
it('TabHeader 与 TabHeaderItem 组合使用', () => {
  clientRectFn!.mockReturnValue({ width: 100, left: 180 });

  const Demo = () => {
    const [selectedIndex, setSelectedIndex] = useState(0);

    return (
      <ThemeProvider theme={defaultTheme}>
        <TabHeader selectedIndex={selectedIndex} onSelect={setSelectedIndex}>
          <TabHeaderItem title="标签1" />
          <TabHeaderItem title="标签2" />
        </TabHeader>
      </ThemeProvider>
    );
  };

  const { getByTestId } = render(<Demo />);

  expect(getByTestId('tab-header-item-0')).toHaveClass(
    'sinoui-tab-label-active',
  );

  act(() => {
    fireEvent.click(getByTestId('tab-header-item-1'));
  });

  expect(getByTestId('tab-header-item-1')).toHaveClass(
    'sinoui-tab-label-active',
  );
});
