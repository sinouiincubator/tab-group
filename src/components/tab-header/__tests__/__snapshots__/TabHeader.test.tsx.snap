// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`全宽模式 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  overflow: hidden;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  font-size: 1rem;
  white-space: nowrap;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  border-bottom: 1px solid rgba(0,0,0,0.12);
}

.c4 {
  position: absolute;
  bottom: 0;
  left: 0;
  -webkit-transition: -webkit-transform 300ms cubic-bezier(0.4,0.0,0.2,1) 0ms,width 300ms cubic-bezier(0.4,0.0,0.2,1) 0ms;
  -webkit-transition: transform 300ms cubic-bezier(0.4,0.0,0.2,1) 0ms,width 300ms cubic-bezier(0.4,0.0,0.2,1) 0ms;
  transition: transform 300ms cubic-bezier(0.4,0.0,0.2,1) 0ms,width 300ms cubic-bezier(0.4,0.0,0.2,1) 0ms;
  height: 2px;
  background-color: #3f51b5;
}

.c1 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  position: relative;
  -webkit-transition: -webkit-transform 0.5s cubic-bezier(0.35,0,0.25,1);
  -webkit-transition: transform 0.5s cubic-bezier(0.35,0,0.25,1);
  transition: transform 0.5s cubic-bezier(0.35,0,0.25,1);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-scrollbar-width: none;
  -moz-scrollbar-width: none;
  -ms-scrollbar-width: none;
  scrollbar-width: none;
  -ms-overflow-style: none;
  overflow: -moz-scrollbars-none;
  overflow-x: scroll;
}

.c1::-webkit-scrollbar {
  display: none;
}

.c1.c1 {
  overflow: hidden;
}

.c1.c1 > .sinoui-tab {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  padding: 10px 8px;
  overflow: hidden;
  display: inline-block;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.c2 {
  height: 48px;
  padding: 0 24px;
  cursor: pointer;
  box-sizing: border-box;
  opacity: 0.6;
  text-align: center;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  white-space: nowrap;
  color: #3f51b5;
  outline: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-letter-spacing: 1.25px;
  -moz-letter-spacing: 1.25px;
  -ms-letter-spacing: 1.25px;
  letter-spacing: 1.25px;
  font-weight: 500;
  position: relative;
  white-space: nowrap;
  -webkit-transition: opacity 0.25s cubic-bezier(0.4,0,0.2,1), color 0.25s cubic-bezier(0.4,0,0.2,1);
  transition: opacity 0.25s cubic-bezier(0.4,0,0.2,1), color 0.25s cubic-bezier(0.4,0,0.2,1);
  -webkit-tap-highlight-color: transparent;
  opacity: 1;
}

.c2:hover {
  opacity: 1;
}

.c3 {
  height: 48px;
  padding: 0 24px;
  cursor: pointer;
  box-sizing: border-box;
  opacity: 0.6;
  text-align: center;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  white-space: nowrap;
  color: rgba(0,0,0,0.87);
  outline: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-letter-spacing: 1.25px;
  -moz-letter-spacing: 1.25px;
  -ms-letter-spacing: 1.25px;
  letter-spacing: 1.25px;
  font-weight: 500;
  position: relative;
  white-space: nowrap;
  -webkit-transition: opacity 0.25s cubic-bezier(0.4,0,0.2,1), color 0.25s cubic-bezier(0.4,0,0.2,1);
  transition: opacity 0.25s cubic-bezier(0.4,0,0.2,1), color 0.25s cubic-bezier(0.4,0,0.2,1);
  -webkit-tap-highlight-color: transparent;
}

.c3:hover {
  opacity: 1;
}

<div
  className="c0 sinoui-tab-header"
  role="tablist"
>
  <div
    className="c1 sinoui-tab-list"
    onScroll={[Function]}
  >
    <div
      aria-selected={true}
      className="c2 sinoui-tab sinoui-tab--active"
      data-testid="tab-header-item-0"
      id="tab-2"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="tab"
      tabIndex={0}
    >
      标签1
    </div>
    <div
      aria-selected={false}
      className="c3 sinoui-tab"
      data-testid="tab-header-item-1"
      id="tab-3"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="tab"
      tabIndex={-1}
    >
      标签2
    </div>
    <div
      className="c4 sinoui-ink-bar"
      data-testid="inkbar"
    />
  </div>
</div>
`;

exports[`正确渲染TabHeader 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  overflow: hidden;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  font-size: 1rem;
  white-space: nowrap;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  border-bottom: 1px solid rgba(0,0,0,0.12);
}

.c2 {
  position: absolute;
  bottom: 0;
  left: 0;
  -webkit-transition: -webkit-transform 300ms cubic-bezier(0.4,0.0,0.2,1) 0ms,width 300ms cubic-bezier(0.4,0.0,0.2,1) 0ms;
  -webkit-transition: transform 300ms cubic-bezier(0.4,0.0,0.2,1) 0ms,width 300ms cubic-bezier(0.4,0.0,0.2,1) 0ms;
  transition: transform 300ms cubic-bezier(0.4,0.0,0.2,1) 0ms,width 300ms cubic-bezier(0.4,0.0,0.2,1) 0ms;
  height: 2px;
  background-color: #3f51b5;
}

.c1 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  position: relative;
  -webkit-transition: -webkit-transform 0.5s cubic-bezier(0.35,0,0.25,1);
  -webkit-transition: transform 0.5s cubic-bezier(0.35,0,0.25,1);
  transition: transform 0.5s cubic-bezier(0.35,0,0.25,1);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-scrollbar-width: none;
  -moz-scrollbar-width: none;
  -ms-scrollbar-width: none;
  scrollbar-width: none;
  -ms-overflow-style: none;
  overflow: -moz-scrollbars-none;
  overflow-x: scroll;
}

.c1::-webkit-scrollbar {
  display: none;
}

<div
  className="c0 sinoui-tab-header"
  role="tablist"
>
  <div
    className="c1 sinoui-tab-list"
    onScroll={[Function]}
  >
    <div
      className="sinoui-tab-label"
    >
      标签内容1
    </div>
    <div
      className="sinoui-tab-label sinoui-tab-label-active"
    >
      标签内容2
    </div>
    <div
      className="c2 sinoui-ink-bar"
      data-testid="inkbar"
    />
  </div>
</div>
`;

exports[`附加内容 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  overflow: hidden;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  font-size: 1rem;
  white-space: nowrap;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  border-bottom: 1px solid rgba(0,0,0,0.12);
}

.c2 {
  position: absolute;
  bottom: 0;
  left: 0;
  -webkit-transition: -webkit-transform 300ms cubic-bezier(0.4,0.0,0.2,1) 0ms,width 300ms cubic-bezier(0.4,0.0,0.2,1) 0ms;
  -webkit-transition: transform 300ms cubic-bezier(0.4,0.0,0.2,1) 0ms,width 300ms cubic-bezier(0.4,0.0,0.2,1) 0ms;
  transition: transform 300ms cubic-bezier(0.4,0.0,0.2,1) 0ms,width 300ms cubic-bezier(0.4,0.0,0.2,1) 0ms;
  height: 2px;
  background-color: #3f51b5;
}

.c3 {
  line-height: 48px;
}

.c1 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  position: relative;
  -webkit-transition: -webkit-transform 0.5s cubic-bezier(0.35,0,0.25,1);
  -webkit-transition: transform 0.5s cubic-bezier(0.35,0,0.25,1);
  transition: transform 0.5s cubic-bezier(0.35,0,0.25,1);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-scrollbar-width: none;
  -moz-scrollbar-width: none;
  -ms-scrollbar-width: none;
  scrollbar-width: none;
  -ms-overflow-style: none;
  overflow: -moz-scrollbars-none;
  overflow-x: scroll;
}

.c1::-webkit-scrollbar {
  display: none;
}

<div
  className="c0 sinoui-tab-header"
  role="tablist"
>
  <div
    className="c1 sinoui-tab-list"
    onScroll={[Function]}
  >
    <div
      className="sinoui-tab-label"
    >
      标签内容1
    </div>
    <div
      className="sinoui-tab-label sinoui-tab-label-active"
    >
      标签内容2
    </div>
    <div
      className="c2 sinoui-ink-bar"
      data-testid="inkbar"
    />
  </div>
  <div
    className="c3 sinoui-tab-header__extra-content"
  >
    <div>
      附加内容
    </div>
  </div>
</div>
`;
